src/core/Akka/IO/TcpConnection.cs

```diff
@@ -46,375 +70,174 @@ namespace Akka.IO
-                    case CloseCommand cmd:
-                        var info = new ConnectionInfo(commander, keepOpenOnPeerClosed: false, useResumeWriting: false);
-                        HandleClose(info, Sender, cmd.Event);
-                        return true;
-                    case ReceiveTimeout _:
-                        // after sending `Register` user should watch this actor to make sure
-                        // it didn't die because of the timeout
-                        Log.Debug("Configured registration timeout of [{0}] expired, stopping", Tcp.Settings.RegisterTimeout);
-                        Context.Stop(Self);
-                        return true;
-                    case WriteCommand write:
-                        // When getting Write before regestered handler, have to buffer writes until registration
-                        var buffered = _writeCommandsQueue.EnqueueSimpleWrites(write, Sender, out var commandSize);
-                        if (!buffered)
-                        {
-                            var writerInfo = new ConnectionInfo(Sender, false, false);
-                            DropWrite(writerInfo, write);
-                        }
-                        else
-                        {
-                            Log.Warning("Received Write command before Register command. " +
-                                        "It will be buffered until Register will be received (buffered write size is {0} bytes)", commandSize);
-                        }
+        private static readonly IOException DroppingWriteBecauseQueueIsFullException =
+            new("Dropping write because queue is full");
 
-                        return true;
-                    default: return false;
-                }
-            };
-        }
+        private int? _partialWriteOffset = null;
 
-        /// <summary>
-        /// Normal connected state.
-        /// </summary>
-        private Receive Connected(ConnectionInfo info)
+        protected TcpConnection(TcpSettings settings, Socket socket, bool pullMode)
         {
-            var handleWrite = HandleWriteMessages(info);
-            return message =>
-            {
-                if (handleWrite(message)) return true;
-                switch (message)
-                {
-                    case SuspendReading _: SuspendReading(); return true;
-                    case ResumeReading _: ResumeReading(); return true;
-                    case SocketReceived _: DoRead(info, null); return true;
-                    case CloseCommand cmd: HandleClose(info, Sender, cmd.Event); return true;
-                    default: return false;
-                }
-            };
-        }
+            Settings = settings;
+            _maxQueuedBytes = settings.WriteCommandsQueueMaxSize; // –1 ⇒ unlimited;
+            _pendingWrites = new Queue<(Write Cmd, IActorRef Sender)>(16);
+            _pullMode = pullMode;
```

Comment:
If `_pullMode` is enabled, which it always is when we're running Akka.Streams.IO.Tcp, then we are only allowed to have 1 `Tcp.Received` in-flight at any given time until we get a `Tcp.ResumeReading`. I'm fine with Akka.Streams backpressuring the socket this way because the buffer settings we added allow for this, whereas doing this with accepting incoming connections is dumb and can result in connection loss.